alchemiscale/storage/statestore.py

```diff
@@ -2878,6 +2962,111 @@ def get_task_restart_patterns(
+
+        if not results:
+            return
+
+        # iterate over all of the results to determine if an applied pattern needs
+        # to be iterated or if the task needs to be cancelled outright
+
+        # Keep track of which task/taskhub pairs would need to be canceled
+        # None => the pair never had a matching restart pattern
+        # True => at least one patterns max_retries was exceeded
+        # False => at least one regex matched, but no pattern max_retries were exceeded
+        cancel_map: defaultdict[Tuple[str, str], Optional[bool]] = defaultdict(
+            lambda: None
+        )
+        to_increment: List[Tuple[str, str]] = []
+        for record in results.records:
+            task_restart_pattern = record["trp"]
+            applies_relationship = record["app"]
+            task = record["task"]
+            taskhub = record["taskhub"]
+            traceback = record["traceback"]
+
+            task_taskhub_tuple = (task["_scoped_key"], taskhub["_scoped_key"])
+
+            # TODO: remove in v1.0.0
+            # tasks that errored, prior to the indtroduction of task restart policies will have no tracebacks in the database
+            if traceback is None:
+                cancel_map[task_taskhub_tuple] = True
+
+            # we have already determined that the task is to be canceled
+            # is only ever truthy when we say a task needs to be canceled
+            if cancel_map[task_taskhub_tuple]:
+                continue
+
+            num_retries = applies_relationship["num_retries"]
+            max_retries = task_restart_pattern["max_retries"]
+            pattern = task_restart_pattern["pattern"]
+            tracebacks: List[str] = traceback["tracebacks"]
+
+            if any([re.search(pattern, message) for message in tracebacks]):
+                if num_retries + 1 > max_retries:
+                    cancel_map[task_taskhub_tuple] = True
+                else:
+                    # to_increment.append(task_taskhub_tuple)
+                    to_increment.append(
+                        (task["_scoped_key"], task_restart_pattern["_scoped_key"])
+                    )
+                    cancel_map[task_taskhub_tuple] = False
+
+        increment_query = """
+        UNWIND $trp_and_task_pairs as pairs
+        WITH pairs[0] as task_scoped_key, pairs[1] as task_restart_pattern_scoped_key
+        MATCH (:Task {`_scoped_key`: task_scoped_key})<-[app:APPLIES]-(:TaskRestartPattern {`_scoped_key`: task_restart_pattern_scoped_key})
+        SET app.num_retries = app.num_retries + 1
+        """
+
+        tx.run(increment_query, trp_and_task_pairs=to_increment)
+
+        # cancel all tasks that didn't trigger any restart patterns (None)
+        # or exceeded a patterns max_retries value (True)
+        for (task, taskhub), _ in filter(
+            lambda values: values[1] is True or values[1] is None, cancel_map.items()
+        ):
+            self.cancel_tasks([task], taskhub, tx=tx)
```

Comment:
Can't add an `else` here since we're looping through and a task/taskhub pair that was marked as False could then later be marked as None, thus cancelling it.

I've grouped tasks to cancel by taskhub.